File: solutions/75-sort-colors/sort-colors.ts
/**
 Do not return anything, modify nums in-place instead.
 */
function sortColors(nums: number[]): void {

    const pivot = 1;

    let left = 0;
    let mid = 0;
    let right = nums.length - 1;
    while (mid <= right) {

        // Checking if value at mid is less than pivot
        if (nums[mid] < pivot) {
            swap(nums, left, mid);
            left += 1;
            mid += 1;

        }
        // Checking if value at mid is greater than pivot
        else if (nums[mid] > pivot) {
            swap(nums, right, mid);
            right -= 1;
            // Edge Case: don't advance the 'mid' pointer if swapped with 'right'!
            // Otherwise, if mid moves after swap, 'one' can remain inthe 'zeroes' partition!
            // mid += 1;
        }
        // otherwise, advance 'mid'
        else {
            mid += 1;
        }

    }

};

function swap(arr: number[], i: number, j: number): void {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

File: solutions/75-sort-colors/sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        num_zeros = 0
        num_ones = 0
        num_twos = 0

        for i in range(0, n):
            curr_num = nums[i]
            if curr_num == 0:
                num_zeros += 1
            elif curr_num == 1:
                num_ones += 1
            elif curr_num == 2:
                num_twos += 1
            else:
                continue
        
        if num_zeros + num_ones + num_twos != n:
            return
        
        index = 0
        while index < n:
            if num_zeros != 0:
                nums[index] = 0
                num_zeros -= 1
            elif num_ones != 0:
                nums[index] = 1
                num_ones -= 1
            else:
                nums[index] = 2
            index += 1
        